### ISSUE TITLE
Rollup does not successfully bundle buildable libraries

### ISSUE THREAD

**Issue by ryanbas21** [nrwl/nx#10395]
## Current Behavior
The @nrwl/web:rollup executor does not bundle buildable libraries that are not in the base package. The bundle will only include files that are in the base-package where the rollup executor is called, and not bundle buildable libraries.

## Expected Behavior
Rollup successfully bundles all required files in the package.

## Steps to Reproduce
github.com/ryanbas21/rollup-issue

### Failure Logs
The build is successful and does build the "dependent" libraries. It does not  however resolve the typescript aliases successfully. See dist/packages in repro and the ESM and UMD build is just the import statements and one function that is included in the base package.

**Comment by NejcZdovc** [comment]
+1 on this one

**Comment by ryanbas21** [comment]
I upgraded the repo to the newest version of nx. I still experience the same issue. it seems to package `libs` as peer deps but only if i build the `libs` first and seperately. there is no automatic detection of libs being a part of the `pacakge` so it errors on `build` if you just try to `build` the package first, alone, without building the libs explicitly.

**Comment by psamusev** [comment]
Same for me, need this funcionality to avoid npm publishing of all buildable libraries ( I need all dependencies from node_modules to consider as external, but include all local workspace projects to the bundle)

**Comment by simondotm** [comment]
yeah, this is problematic, since the whole benefit of Nx workspaces with libraries is that we want to be able to bundle them in our apps.

I've tried forcing my own `rollup.config.cjs`, which does at least bundle local libs now, but still not a solution because the `@nrwl/rollup:rollup` executor writes the output `package.json` as the last step, which only contains external deps from the entry module, not the bundled output.

Here's my wip config in case its helpful to anyone:
```
// https://github.com/rollup/plugins/tree/master/packages/typescript
const typescript = require('@rollup/plugin-typescript').default;
const generatePackageJson = require('rollup-plugin-generate-package-json')

module.exports = (config, context) => {
  const filteredPlugins = []
  for (const plugin of config.plugins) {
      console.log(JSON.stringify(plugin, null, 3))
      if ([
          'peer-deps-external', 
          'rollup-plugin-nx-analyzer',
          'commonjs'
          // 'node-resolve'
      ].includes(plugin.name)) {
          filteredPlugins.push(plugin)
      }
  }
  const updatedConfig = {
    // ...config,
    input: config.input,
    output: {
        ...config.output,
        sourcemap: true,
    },
    plugins: [
        ...filteredPlugins,
        typescript({ tsconfig: context.tsConfig, compilerOptions: { outDir: config.output.dir, sourceMap: true } }), 
        generatePackageJson()
    ]
  }
  return updatedConfig
}
```

I'm using the `typescript` plugin and so far that seems to magically pick up the library aliases in my `tsconfig.base.json` and do the right thing.

**Comment by simondotm** [comment]
A little more investigation reveals a number of things in the `@nrwl/rollup:rollup` implementation that are causing issues for my node app use of this plugin (my aim: I have a node app and I want to bundle my app entry code and any `@myorg/...` type workspace libraries it imports):

1. [This line](https://github.com/nrwl/nx/blob/fc2ea3c88a1e64882249171285f780b08f0189f0/packages/rollup/src/executors/rollup/rollup.impl.ts#L222) sets a tsconfig override that sets `rootDir` to the project root, but then also passes in the `paths` aliases from the `tsconfig.base.json` which are now relative to `rootDir` and therefore never resolve as an internal module that can be bundled
2. [This line](https://github.com/nrwl/nx/blob/fc2ea3c88a1e64882249171285f780b08f0189f0/packages/rollup/src/executors/rollup/rollup.impl.ts#L50) only asks for a shallow dependency graph (`shallow=true`), so it only returns local & npm deps it finds in the top level project 
3. [This line](https://github.com/nrwl/nx/blob/fc2ea3c88a1e64882249171285f780b08f0189f0/packages/rollup/src/executors/rollup/rollup.impl.ts#L244) adds `@rollup/plugin-node-resolve` plugin, which I find is ridiculously aggressive at including external and even node builtins like "fs" in the bundle - ideally we'd have the Nx option to not use this plugin if desired
4. [This line](https://github.com/nrwl/nx/blob/fc2ea3c88a1e64882249171285f780b08f0189f0/packages/rollup/src/executors/rollup/rollup.impl.ts#L164) unconditionally generates an output `package.json`, but it's flawed because a) the external deps do not match the bundleas per (2) and b) the entry point names are wrong - ideally we'd have the `generatePackageJson` Nx option here to disable it and use a rollup plugin like `rollup-plugin-generate-package-json` instead, or fix the Nx plugin output to contain deps that are actually in the output bundle

My workaround for now:
* Use my own rollup config that ignores most of the Nx rollup config
* Patch `rollup.impl.js` to not write `package.json`
* Set my own `config.plugins`
* Set my own `config.external` matcher function

```
const generatePackageJson = require('rollup-plugin-generate-package-json')
const commonjs = require('@rollup/plugin-commonjs')
const sizes = require("rollup-plugin-sizes")
const typescript2 = require('rollup-plugin-typescript2')

module.exports = (config, context) => {

    function isExternal(moduleName) {
        const internal = moduleName.includes('@myorg') || moduleName.includes('myorg/apps') || moduleName.includes('myorg/libs') || /^\.{0,2}\//.test(moduleName)
        return !internal
    }

    return {
        input: config.input,
        output: {
            ...config.output,
            sourcemap: true,
            intro,
            outro,
        },
        external: isExternal, 
        plugins: [
            typescript2({ tsconfig: context.tsConfig, compilerOptions: { outDir: config.output.dir, sourceMap: true } }),
            commonjs(),
            generatePackageJson({
                baseContents: (pkg) => ({
                    name: context.projectRoot.split('/').slice(-1)[0],
                    main: "./main.js",
                    dependencies: {},
                    private: true,
                    type: "module"
                })
            }),
            sizes({ details: true })
        ]
    }
```

**Comment by simondotm** [comment]
Here's my hacky patcher that I call in my `rollup.config.cjs` file:
```
const fs = require('fs')

function patchNxPlugin() {
  const target = './node_modules/@nrwl/rollup/src/executors/rollup/lib/update-package-json.js'
  const data = fs.readFileSync(target, 'utf-8');

  const patchMatch = 'const hasEsmFormat'
  const patchReplace = `return;${patchMatch}`
  if (!data.includes(patchReplace)) {
    const output = data.replace(patchMatch, patchReplace);
    fs.writeFileSync(target, output, 'utf-8');
    throw new Error(`PATCHED @nrwl/rollup updatePackageJson, re-run build`)
  }else{
    console.log(`ALREADY PATCHED @nrwl/rollup updatePackageJson`)
  }
}
```

**Comment by simondotm** [comment]
This is still an issue in 16.1.1
For me, just patching this line from `true` to `false` to not use shallow dependency generation seems to solve most of the problem, since it now evaluates dependencies from not just the entry module, but the whole project.
https://github.com/nrwl/nx/blob/1488728aa3e848156212aed9f9e6994bae08590e/packages/rollup/src/executors/rollup/rollup.impl.ts#L61

That said, it still emits dependencies that might be optimized out by rollup tree shaking.

Would prefer an option to completely prevent Nx writing the `package.json` at all, because even though we now have `        "updateBuildableProjectDepsInPackageJson": false` option, it still writes a stub `package.json` that overwrites the one generated by my custom rollup config.

**Comment by bigcakes** [comment]
Same issue here and a discussion was started as well: https://github.com/nrwl/nx/discussions/17605

**Comment by arielweinberger** [comment]
Bumping this. Can anyone at Nx please consider the suggestion above? It provide a temporary solution for many.

**Comment by jaysoo** [comment]
We'll take a look.

**Comment by jahumes** [comment]
@jaysoo Any update on this? We have a new library we are trying to publish, but it isn't pulling in other buildable libraries from inside our repo.

**Comment by igorkamyshev** [comment]
lol 😂 

Maybe it's better to fix an issue rather than mark it as stale and close?

**Comment by antoniojps** [comment]
Been battling this for a few days... beats the purpose of using NX with publishable libs.

**Comment by jaysoo** [comment]
Bundling isn't hard, but rolling up `d.ts` files takes more work. I have a draft open for this: https://github.com/nrwl/nx/pull/27549

**Comment by rachirib** [comment]
This is still not working on `19.x` is any help on how this can done ?
